File: src/main/java/com/heybys/oddments/security/oauth2/user/OAuth2UserInfoFactory.java
package com.heybys.oddments.security.oauth2.user;

import java.util.Map;

import com.heybys.oddments.exception.OAuth2AuthenticationProcessingException;
import com.heybys.oddments.fooddelivery.domain.generic.AuthProvider;

public class OAuth2UserInfoFactory {

    private OAuth2UserInfoFactory() {}

    public static OAuth2UserInfo getOAuth2UserInfo(String registrationId, Map<String, Object> attributes) {
        if (registrationId.equalsIgnoreCase(AuthProvider.GOOGLE.toString())) {
            return new GoogleOAuth2UserInfo(attributes);
        } else if (registrationId.equalsIgnoreCase(AuthProvider.FACEBOOK.toString())) {
            return new FacebookOAuth2UserInfo(attributes);
        } else if (registrationId.equalsIgnoreCase(AuthProvider.GITHUB.toString())) {
            return new GithubOAuth2UserInfo(attributes);
        } else {
            throw new OAuth2AuthenticationProcessingException(
                    "Sorry! Login with " + registrationId + " is not supported yet.");
        }
    }
}

File: src/main/kotlin/com/heybys/oddments/security/oauth2/user/OAuth2UserInfoFactory.kt
package com.heybys.oddments.security.oauth2.user

import com.heybys.oddments.exception.OAuth2AuthenticationProcessingException
import com.heybys.oddments.fooddelivery.domain.generic.AuthProvider

object OAuth2UserInfoFactory {
    fun getOAuth2UserInfo(registrationId: String, attributes: Map<String, Any>): OAuth2UserInfo {
        return when (registrationId.uppercase()) {
            AuthProvider.GOOGLE.toString() -> GoogleOAuth2UserInfo(attributes)
            AuthProvider.FACEBOOK.toString() -> FacebookOAuth2UserInfo(attributes)
            AuthProvider.GITHUB.toString() -> GithubOAuth2UserInfo(attributes)
            else ->
                throw OAuth2AuthenticationProcessingException(
                    "Sorry! Login with $registrationId is not supported yet.",
                )
        }
    }
}
